Add PGplot, a method for generating graphs.

somiaj opened the pull request:
Here is something I've been working on for a while, but probably best to start getting some feedback.

First, PGplot is a new method for generating dynamic graphs and is meant to be a replacement of WWPlot and PGgraphmacros.pl.  The first main difference is PGplot is just an API to store data that can then be used to generate a graph using multiple outputs. Right now only TikZ pgfplots and GD (for testing mostly) is supported. I started trying to fully support both, but as I added more features, I focused only on TikZ pgfplots features (so GD only supports simple things and should probably be removed, but leaving it here for now).

In addition there is a new macro PGplotmacros.pl that is meant to be a drop in replacement of PGgraphmacros.pl. This works in most cases provided the problem doesn't call any GD features directly and doesn't use filling (since I couldn't figure out how to get the GD fill method to work with TikZ).  My hope was this could just replace PGgraphmacros.pl, but there are a non trivial number of problems in the OPL that call GD features directly, that this only worked on most of the OPL problems I tested.

Here are a set of examples and testing problems I have been working with to try out. This contains the problems and a few set definition files you can import. [PGplot-examples.tar.gz](https://github.com/openwebwork/pg/files/13462854/PGplot-examples.tar.gz)

I am starting this as a draft, as I suspect there are going to be lots of changes as we think about the PGplot API being used and what sort of features it should support.

somiaj commented:
Well looks like my method of splittings up in the `macros/` directory is causing a lot of tests to fail, since it cannot find the objects that are needed. I will have to look into that.

drgrice1 reviewed (commented):
Here are some initial thoughts on this.

The `PGplotmacros.pl` needs to be renamed to something that does not contain the word "macros".

somiaj commented on macros/graph/unionImage.pl:
```diff
@@ -66,12 +66,12 @@ sub Image {
-	$image = alias(insertGraph($image)) if (ref($image) eq "WWPlot");
-	$image = alias($image) unless ($image =~ m!^(/|https?:)!i);         # see note
+	$image = alias(insertGraph($image)) if (ref($image) eq "WWPlot" || ref($image) eq 'PGplot');
+	$image = alias($image) unless ($image =~ m!^(/|https?:)!i);                                    # see note
 
 	if ($ilink) {
 		$ilink = alias(insertGraph($ilink)) if (ref($ilink) eq "WWPlot");
-		$ilink = alias($ilink) unless ($ilink =~ m!^(/|https?:)!i);         # see note
+		$ilink = alias($ilink) unless ($ilink =~ m!^(/|https?:)!i);                                # see note
```
Thanks, I missed that location.

somiaj commented on t/macros/load_macros.t:
```diff
@@ -34,7 +34,17 @@ my %baseMacros = (
+my %brokenMacros = (
+	'answerDiscussion.pl' => 1,
+	'PG_CAPAmacros.pl'    => 1,
+	'Axes.pl'             => 1,
+	'GD.pl'               => 1,
+	'Tikz.pl'             => 1,
+	'Data.pl'             => 1,
+	'PGplotmacros.pl'     => 1,
```
Would modules in `lib` be preferable over combining `Axes.pl`, etc in a single file? I just feel it would be better to have everything loaded from a macro as needed, but it probably doesn't matter.

I'll push `PGplotmacros.pl` to a second PR. The current way it is setup, it modifies the WWplot methods, causing other macros to fail their tests, and PGplot can be tested without this wrapper.

somiaj commented:
I removed the `PGplotmacros.pl` wrapper, and will add that to a future PR. Also combined everything into a single macro file, so now this is just a new macro for creating graphs. The primary output at the time is PGFPlots via a LatexImage, though it is meant to be able to support different outputs.

I'm also noticing a difference in perltidy here and in the tests. In the failed perltidy test, the test wants to add an additional tab to the list of or statements I modified in `PGbasicmacros.pl`, but perltidy doesn't want that tab here. I'm running perltidy version `20220613` in Debian bookworm, could there be a change there?

drgrice1 commented:
Rebase onto develop now, and it will fix the perltidy issue.

somiaj commented:
I still have some difference between my local perltidy and the test results. My guess is the older version, so I'll look into updating that.

drgrice1 commented:
Make sure you specifically use Perl::Tidy version 20220613.  Both newer and older versions will give different results.  This is one of my annoyances with Perl::Tidy.  When they release new versions, the break compatibility with older versions.

somiaj commented:
That is strange, I am using 20220613. I did perltidy `PGbasicmacros.pl`, and last night when I tried to perltidy that file again, I had no changes. I just tried, and now I have the same changes as the workflow test. Unsure what is going on, but the issue seems to have gone away.

drgrice1 commented on macros/graph/PGplot.pl:
```diff
@@ -0,0 +1,1914 @@
+	my $py   = $ux;
+	my $hbx  = $x2 - 7 * $w * $ux;
+	my $hby  = $y2 - 7 * $w * $uy;
+	my $head = new GD::Polygon;
```
This should be
```suggestion
	my $head = GD::Polygon->new;
```

drgrice1 commented on macros/graph/PGplot.pl:
```diff
@@ -0,0 +1,1914 @@
+
+=head1 USAGE
+
+The axes object should be accessed through a PGplot object using C<$plot-E<gt>axes>.
```
Instead of using the `E<gt>` syntax, make this readable directly with `C<< $plot->axes >>`.  Not the spaces in that are required.   I saw this several other places in this file where this should be fixed as well.

drgrice1 commented on macros/graph/PGplot.pl:
```diff
@@ -0,0 +1,1914 @@
+
+sub new {
+	my $class = shift;
+	my $self  = {
+		xaxis  => {},
+		yaxis  => {},
+		styles => {
+			title      => '',
+			grid_color => 'gray',
+			grid_style => 'solid',
+			grid_alpha => 40,
+			show_grid  => 1,
+		},
+		@_
+	};
+
+	bless $self, $class;
```
Here again, bless `$self` directly instead of on a separate line.

drgrice1 commented on macros/graph/PGplot.pl:
```diff
@@ -0,0 +1,1914 @@
+package PGplot::Data;
+
+sub new {
+	my $class = shift;
```
Unpack the arguments here by adding `%options`, and bless `$self` below directly.

drgrice1 commented on macros/graph/PGplot.pl:
```diff
@@ -0,0 +1,1914 @@
+}
+
+sub set_function {
+	my $self = shift;
```
Unpack arguments.

drgrice1 commented on macros/graph/PGplot.pl:
```diff
@@ -0,0 +1,1914 @@
+}
+
+sub add {
+	my $self = shift;
```
One last place where arugments need to be unpacked.

drgrice1 reviewed (commented):
I think this is a good start.  I noted some code issues.

You have several examples in the tar file that you provided that reference a `PGplotmacros.pl` macro which does not exist.  I would like to see that actually work (but not with a `PGplotmacros.pl` file).  That should be the main goal of this.  That is to make it so that the code in `PGgraphmacros.pl` can be eventually replaced with calls to methods here and it will work for existing problems that use that macro.  There is the issue of problems using `GD` objects directly, but I am sure we can find a way to work around that.

I don't like the merge of the previous separate files all into this one file, making it almost 2000 lines long.  Instead the `PGplot`, `PGplot::Tikz`, `PGplot::GD`, and `PGPlot::Axes` should be made actual modules (each a separate file).  Then this macro should have just the basics need to instantiate the object (much like how the `PGtikz.pl` and `PGlateximage.pl` macros work with respect to the `LaTeXImage` package).  The documentation on how to use the macro should stay here for problem authors (again like the above mentioned macros).  Modular code is much better than a single massive file.

somiaj commented:
@drgrice1 Thanks for the review. @pstaabp has been doing a bit of work on this and has split everything back out and moved this to `lib/`. He is focused on getting this to work with jsxgraph for html, and also have a way to just export JSON from to plot data. I was informed he may add a PR to my branch with the changes, when those come in I'll implement any of these which are still relevant.

somiaj commented:
I removed the `PGplotmacros.pl` for now because I thought it would be easier to focus on getting this to an acceptable state before dealing with trying to make existing problems work with a wrapper. I was hoping initially to have a drop in replacement, but as I kept running into how many problems were calling GD directly, that I saw this was going to take a lot more work and should probably be dealt with separately. I also think some of my approach, which was working for some problems, was overriding things in an incorrect way, which is what caused it to break all the auto tests.

I do agree it should be a goal that we can just have a drop in replacement and all current problems are updated to using something newer.

somiaj commented:
I updated this, I removed the change of putting everything in a single file, then added @pstaabp change to move everything to `lib/`, and some code clean up. I also updated all of @drgrice1 suggestions as well.

This move changed the package to just `Plots`, and the command to create a `Plots` object to `Plot()`. Since the `Plots` object is now apart of `lib/`, the following patch https://github.com/somiaj/webwork2/tree/pgplot-libs is needed. I have updated all the examples naming.

[Plots_examples.tar.gz](https://github.com/user-attachments/files/16651201/Plots_examples.tar.gz)

In updating this I did notice an issue with `pdf2svg` and `xelatex` set as the default. It seems that `xelatex` isn't fully compatible with `pgfplots`, though switching this to use `dvisvgm` by default works just fine (since that calls `latex` directly). The issue I noticed was with the marks test (Problem 8 of the Test set), the rest worked just fine. I think one of the marks uses a symbol that conflicts with `xelatex`. I am just documenting this issue for now, the default setup using `dvisvgm` works properly.

somiaj commented:
Add the very basics of JSXGraph and Plotly.js output support. Examples updated.

[Plots_examples.tar.gz](https://github.com/user-attachments/files/17691181/Plots_examples.tar.gz)

somiaj commented:
Thinking about adding JSXGraph and Plotly.js output method a bit more, I think this approach is not the best. Plots is mostly designed around creating non dynamic graphs, so at best adding javascript output could do is allow some moving around/zooming in/out of the graph in some situations.

I think a better approach would create a JSXGraph and Plotly.js plots object that is a subclass, built on the same `Plots::Data` objects, so it can fall back to using TikZ (or some other non dynamic output) for hard copy, instead of try to make Plots do both non-dynamic and dynamic graphs.

drgrice1 commented:
JSXGraph does a great job of rendering static graphs.  I use it for that purpose in problems via a local macro that I have.

somiaj commented:
I was more thinking about also wanting to be able to also use it for dynamic graphs from the same macro. And unsure if adding the dynamic side of things should be done in Plots, or better to just have a separate macro for that. But so far it is working well with the setup I have here.

drgrice1 commented:
Yeah, I think that this is really for static graphs (dynamically generated but not interactive).  I mean you could have dynamic in the sense of pan and zoom I suppose (both Plotly and JSXGraph support that).

somiaj commented:
Agreed, so my thought was Plotly and JSXGraph could be separate macros that build on top of this, so they can create static and dynamic graphs (as opposed to what I'm working on now which is just making it an optional output method). Though I guess they can split off later.

somiaj commented:
Updated everything and rebased against `PG-2.20`. I ditched the idea of support poltly at the time, and have mostly worked on JSXgraph, with the idea that JSXgraph should be the default HTML output and TikZ the hardcopy (though I do like TikZ graphs, I think they generate much slower).

Updated sample problems, think I have JSXgraph mostly in line with TikZ. TikZ does support some things JSXgraph doesn't though the only real thing that I didn't find a good solution was filling area between curves, seems this isn't default in JSXgraph, also if the time comes to try to replace PGplotmacros with this, being able to fill a region given a point is also needed.

[Plots_examples.tar.gz](https://github.com/user-attachments/files/20560271/Plots_examples.tar.gz)

pstaabp commented on macros/graph/plots.pl:
```diff
@@ -0,0 +1,419 @@
+# FOR A PARTICULAR PURPOSE.  See either the GNU General Public License or the
+# Artistic License for more details.
+################################################################################
+
```
We've gotten rid of this header material.

somiaj replied:
Thanks. It is now gone.

somiaj commented:
Vector field support added. Examples updated.

[Plots_examples.tar.gz](https://github.com/user-attachments/files/20649653/Plots_examples.tar.gz)

drgrice1 commented:
@somiaj: I added a pull request with the background/border for TikZ images that I mentioned.

somiaj commented:
@drgrice1 thanks. If you get a chance, can you set it up so a modal can be used to zoom in on the JSXGraph output if the option `$plot->axes->style('jsx_navigation')` is not set?

drgrice1 commented:
I added a pull request that cleans up the `addToPreamble` call.  That was quite unreadable.

drgrice1 commented:
I added another pull request with the `imageview` dialog when the `jsx_navigation` option is false.  There are also some other changes that are needed.

drgrice1 commented:
I added another pull request to this branch.  I figured out what was really going on with the image in the dialog and fixed it.

somiaj commented:
Grr, I thought I was on my branch with your new merge....should I revert it or do a force push to revert?
